### PR TITLE
[Graph Browser 2] Fix/Improve opening observation node page from observation chart

### PR DESCRIPTION
**Pull request by chejennifer**
- update the flask server endpoint to use new statVarObservation table & return a mapping of date to dcids
- update observationChart component to retrieve observation dcids for all the data points on the chart & have it in the state so onClick will open the observation node page right away

### CODE REVIEW

File: static/js/browser2/observation_chart.tsx [hunk 1]
```diff
@@ -96,6 +109,26 @@ export class ObservationChart extends React.Component<
       .on("mouseleave", this.handleDotLeave(svgContainerId));
   }
 
+  private fetchObservationDcidsData(): void {
+    let request = `/api/browser/observation-ids?place=${this.props.placeDcid}&statVar=${this.props.statVarId}`;
+    if (this.props.sourceSeries.measurementMethod) {
+      request =
+        request +
+        `&measurementMethod=${this.props.sourceSeries.measurementMethod}`;
```
Comment (beets) [inline static/js/browser2/observation_chart.tsx:117]: nit: request could be folded in too:

`${request}&measurementMethod...`

File: static/js/browser2/observation_chart.tsx [hunk 2]
```diff
@@ -143,22 +176,11 @@ export class ObservationChart extends React.Component<
...
-        request + `&obsPeriod=${this.props.sourceSeries.observationPeriod}`;
-    }
-    const obsDcidPromise = axios.get(request).then((resp) => resp.data);
-    obsDcidPromise.then((obsDcid) => {
+    const obsDcid = this.state.obsDcidMapping[date];
+    if (obsDcid) {
       const uri = URI_PREFIX + obsDcid;
       window.open(uri);
-    });
+    }
```
Comment (beets) [inline static/js/browser2/observation_chart.tsx:184]: should there be a response in the else case (error message perhaps?)

**Review by beets — APPROVED** [review]

File: static/js/browser2/observation_chart.tsx (hunk 2 above)
Comment (chejennifer) [reply static/js/browser2/observation_chart.tsx:184]: I think in the original browser, the response was to take the user to the landing page of the browser tool. I do think an error message might be more useful, so I can try to do that

Comment (chejennifer) [reply static/js/browser2/observation_chart.tsx:193]: added a simple error message: https://screenshot.googleplex.com/3RHNa8mwk4Zy6Fy

File: static/js/browser2/observation_chart.tsx (hunk 1 above)
Comment (chejennifer) [reply static/js/browser2/observation_chart.tsx]: done

File: server/routes/api/browser.py [hunk 3]
```diff
@@ -78,36 +78,35 @@ def get_observation_id():
...
+            ?svObservation typeOf StatVarObservation .
+            ?svObservation variableMeasured {} . 
+            ?svObservation observationAbout {} .
+            ?svObservation dcid ?dcid .
+            ?svObservation observationDate ?obsDate .
             {}
             {}
         }}
-    '''.format(stat_var_id, place_id, date, measurement_method_line,
+    '''.format(stat_var_id, place_id, measurement_method_line,
```
Comment (shifucun) [inline server/routes/api/browser.py:100]: nit: can change *_line to *_triple

File: server/routes/api/browser.py [hunk 4]
```diff
@@ -78,36 +78,35 @@ def get_observation_id():
...
     result = ""
     (_, rows) = dc.query(sparql_query)
-    if len(rows) > 0:
-        cells = rows[0].get("cells", [])
-        if len(cells) > 0:
-            result = cells[0].get('value', '')
+    result = {}
+    for row in rows:
+        cells = row.get('cells', [])
+        if not len(cells) == 2:
```
Comment (shifucun) [inline server/routes/api/browser.py:107]: nit: if len(cells) != 2

File: server/tests/api_browser_test.py [hunk 5]
```diff
@@ -24,71 +24,84 @@ class TestObservationId(unittest.TestCase):
...
+            ?svObservation observationDate ?obsDate .
             
             
         }
     '''
+        obs_date = "2001"
         expected_obs_id = "test_obs_id"
 
         def side_effect(query):
+            print(query)
```
Comment (shifucun) [inline server/tests/api_browser_test.py:52]: delete print?

File: static/js/browser2/observation_chart.tsx [hunk 6]
```diff
@@ -33,32 +33,54 @@ const HEIGHT = 250;
...
   sourceSeries: SourceSeries;
   idx: number;
   statVarId: string;
   placeDcid: string;
+  hasClickableDots: boolean;
+}
+
+interface ObservationChartStateType {
   canClickDots: boolean;
+  obsDcidMapping: { [date: string]: string };
```
Comment (shifucun) [inline static/js/browser2/observation_chart.tsx:49]: obsDcidMapping -> dateToDcid

File: server/routes/api/browser.py (hunk 3 above)
Comment (chejennifer) [reply server/routes/api/browser.py]: done

File: static/js/browser2/observation_chart.tsx (hunk 6 above)
Comment (chejennifer) [reply static/js/browser2/observation_chart.tsx]: done

File: server/tests/api_browser_test.py (hunk 5 above)
Comment (chejennifer) [reply server/tests/api_browser_test.py]: done

File: server/routes/api/browser.py (hunk 4 above)
Comment (chejennifer) [reply server/routes/api/browser.py]: done

**Review by shifucun — APPROVED** [review]

**Pull request merged**